packages/gatsby-cli/src/reporter/reporter.ts

```diff
@@ -0,0 +1,267 @@
+    const reporterError = reporter.error(errorMeta, error)
+    trackError(`GENERAL_PANIC`, { error: reporterError })
+    prematureEnd()
+    process.exit(1)
+  }
+
+  panicOnBuild(
+    errorMeta: string | ErrorMeta | Error,
+    error?: Error | Error[]
+  ): IStructuredError | IStructuredError[] {
+    const reporterError = reporter.error(errorMeta, error)
+    trackError(`BUILD_PANIC`, { error: reporterError })
+    if (process.env.gatsby_executing_command === `build`) {
+      prematureEnd()
+      process.exit(1)
+    }
+    return reporterError
+  }
+
+  // TODO: I wish this could also be typed better,
+  // but our usages are sooo across the board right now
+  error(
+    errorMeta: string | ErrorMeta | Error,
+    error?: Error | Error[]
+  ): IStructuredError | IStructuredError[] {
+    let details: {
+      error?: Error
+      context: {}
+    } = {
+      error: undefined,
+      context: {},
+    }
+
+    // Many paths to retain backcompat :scream:
+    // 1.
+    //   reporter.error(any, Error);
+    //   reporter.error(any, [Error]);
+    if (error) {
+      if (Array.isArray(error)) {
+        return error.map(errorItem =>
+          this.error(errorMeta, errorItem)
+        ) as IStructuredError[]
+      }
+      details.error = error
+      details.context = {
+        sourceMessage: errorMeta + ` ` + error.message,
+      }
+      // 2.
+      //    reporter.error(Error);
+    } else if (errorMeta instanceof Error) {
+      details.error = errorMeta
+      details.context = {
+        sourceMessage: errorMeta.message,
+      }
+      // 3.
+      //    reporter.error([Error]);
+    } else if (Array.isArray(errorMeta)) {
+      // when we get an array of messages, call this function once for each error
+      return errorMeta.map(errorItem =>
+        this.error(errorItem)
+      ) as IStructuredError[]
+      // 4.
+      //    reporter.error(errorMeta);
+    } else if (typeof errorMeta === `object`) {
+      details = Object.assign({}, errorMeta as ErrorMeta)
+      // 5.
+      //    reporter.error('foo');
+    } else if (typeof errorMeta === `string`) {
+      details.context = {
+        sourceMessage: errorMeta,
+      }
+    }
+
+    const structuredError = constructError({ details })
+    if (structuredError) {
+      reporterActions.createLog(structuredError)
+    }
+
+    // TODO: remove this once Error component can render this info
+    // log formatted stacktrace
+    if (structuredError.error) {
+      this.log(errorFormatter.render(structuredError.error))
+    }
+    return structuredError
+  }
+
+  /**
+   * Set prefix on uptime.
+   */
+  uptime(prefix: string): void {
+    this.verbose(`${prefix}: ${(process.uptime() * 1000).toFixed(3)}ms`)
+  }
+
+  verbose(text: string): void {
+    if (isVerbose) {
+      reporterActions.createLog({
+        level: LogLevels.Debug,
+        text,
+      })
+    }
+  }
+
+  success = (text: string) =>
```

Comment:
Type the return values of these. It can't be inferred, as `reporterActions` is untyped